ard_controller/ard_controller.ino

```diff
@@ -15,45 +25,274 @@ struct JointAngleStruct {
+        calculated_checksum += received_data;
+        data_byte_counter--;
+        if (data_byte_counter == 0) {
+          data_byte_counter = DATA_BYTE_LENGTH;
+          sr_state = READ_R_SHOULDER;
+        }
+        break;
+      case READ_R_SHOULDER:
+        temporary_packet_data.right_shoulder += received_data;
+        calculated_checksum += received_data;
+        data_byte_counter--;
+        if (data_byte_counter == 0) {
+          data_byte_counter = DATA_BYTE_LENGTH;
+          sr_state = READ_R_ELBOW;
+        }
+        break;
+      case READ_R_ELBOW:
+        temporary_packet_data.right_elbow += received_data;
+        calculated_checksum += received_data;
+        data_byte_counter--;
+        if (data_byte_counter == 0) {
+          data_byte_counter = DATA_BYTE_LENGTH;
+          sr_state = READ_CHECKSUM;
+        }
+        break;
+      case READ_CHECKSUM:
+        temporary_packet_data.checksum = received_data;
+
+        // calculated checksum must match received checksum; otherwise there's an error
+        if (temporary_packet_data.checksum == 0xFF - (calculated_checksum % 256)) {
+          // good packet
+          calculated_checksum = 0x00;
+          preamble_counter = PREAMBLE_LENGTH;
+
+          temporary_packet_data.checksum_error = false;
+          temporary_packet_data.packet_available = true;
+
+          validated_packet_data.left_hip = temporary_packet_data.left_hip;
+          validated_packet_data.left_knee = temporary_packet_data.left_knee;
+          validated_packet_data.right_hip = temporary_packet_data.right_hip;
+          validated_packet_data.right_knee = temporary_packet_data.right_knee;
+          validated_packet_data.left_shoulder = temporary_packet_data.left_shoulder;
+          validated_packet_data.left_elbow = temporary_packet_data.left_elbow;
+          validated_packet_data.right_shoulder = temporary_packet_data.right_shoulder;
+          validated_packet_data.right_elbow = temporary_packet_data.right_elbow;
+          validated_packet_data.checksum = temporary_packet_data.checksum;
+          validated_packet_data.checksum_error = temporary_packet_data.checksum_error;
+          validated_packet_data.packet_available = temporary_packet_data.packet_available;
+          validated_packet_data.data_request = temporary_packet_data.data_request;
+
+          sr_state = INIT;
+        }
+        else {
+          temporary_packet_data.checksum_error = true;
+          temporary_packet_data.packet_available = false;
+          sr_state = INIT;
+        }
+        break;
+      default:
+        sr_state = INIT;
+        break;
+      }
   }
 }
 
+void sensor_data_response() {
+  int raw_sum = 0;
+  int checksum;
+  long angle, goal;
+  int b1, b2;
+
+  // send preamble
+  for (int i = 0; i < PREAMBLE_LENGTH; i++) {
+    Serial1.write((byte)255);
+  }
+
+  // send imu and joint angle data
+  for (int i = 0; i < NUM_JOINTS + NUM_IMU_AXES; i++) {
+    angle = random(360);
```

Comment:
Why is there a random here? I am assuming for testing?
If that is the case, can you replace it with a method that can have this random implementation?
I don't think it's a good idea to test within the core functionality methods. Let's abstract out the details and test it that way

Reply:
Yea it's there as a placeholder until the ODrive is setup with the Arduino. I don't agree with spending the time to abstract something out that's going to be deleted/replaced anyway. It's not like we're going to forget that those are there.

Reply:
You would need a getAngle method or something like that when you actually put in real code right?
So we could put it now and just add this random implementation inside that.
I'll leave this up to your judgment

Reply:
For now, since I don't know what the implementation of the ODrive code is gonna look like yet, I'll say let's leave it as is. This is easy to replace.

Reply:
Got it